crates/hir_ty/src/infer/pat.rs

```diff
@@ -232,16 +233,33 @@ impl<'a> InferenceContext<'a> {
                     self.infer_pat(pat_id, &elem_ty, default_bm);
                 }
 
-                let pat_ty = match expected.kind(Interner) {
-                    TyKind::Array(_, const_) => TyKind::Array(elem_ty, const_.clone()),
-                    _ => TyKind::Slice(elem_ty),
-                }
-                .intern(Interner);
                 if let &Some(slice_pat_id) = slice {
-                    self.infer_pat(slice_pat_id, &pat_ty, default_bm);
+                    let rest_pat_ty = match expected.kind(Interner) {
+                        TyKind::Array(_, length) => match length.data(Interner).value {
+                            ConstValue::Concrete(length) => match length.interned {
+                                ConstScalar::Usize(length) => {
+                                    let length =
+                                        length.checked_sub((prefix.len() + suffix.len()) as u64);
+                                    TyKind::Array(
+                                        elem_ty.clone(),
+                                        crate::consteval::usize_const(length),
+                                    )
+                                    .intern(Interner)
+                                }
+                                ConstScalar::Unknown => self.err_ty(),
```

Comment:
this should not result in err_ty, rather it should just reuse the length value as before.

Reply:
Made them return an unresolved constant (but array of the correct type), because it's a type error.